### ISSUE TITLE
L043 - sqlfluff fix returns broken coalesce statement. 

### ISSUE THREAD

**Issue by GClunies** [sqlfluff/sqlfluff#972]
## Expected Behaviour
When `sqlfluff fix` applies a fix based on [L043](https://docs.sqlfluff.com/en/latest/rules.html?highlight=ignore#sqlfluff.core.rules.Rule_L043), the end result should be something like:
![image](https://user-images.githubusercontent.com/28986302/115073403-24de2080-9ead-11eb-8413-76e456a62669.png)

## Observed Behaviour
Instead, for some of the `case` statements that are fixed by `L043` sqlfluff returns, the returned fix is missing the characters `coalesce(` needed for the coalesce to work. 
![image](https://user-images.githubusercontent.com/28986302/115073366-155ed780-9ead-11eb-832c-796721912aaf.png)

Other times, the case statement (in another part of the model or in another model entirely) is fixed as expected _including_ the `coalesce(` at the start. Not sure why this would be the case? 🤔 

If I run `sqlfluff fix` on the same model where I observe this strange behaviour, I continue to get the same missing `coalesce(` in the returned fix.

## Steps to Reproduce

I have called out where the L043 fix is producing an unexpected fix with the comments `-- BEFORE L043 FIX` and `--AFTER L043 FIX` in the example SQL files below.

<details>
<summary>Example dbt model BEFORE running "sqlfluff fix"</summary>
<p>

```sql
-- filename: sl_stripe_subscriptions_with_perks_history.sql

{{
  config(
    materialized='table',
    sort=['user_id', 'created']
  )
}}

with subscriptions as (

    select * from {{ ref('stg_sl_stripe_subscriptions') }}

),

customers as (

    select * from {{ ref('stg_sl_stripe_customers') }}

),

perks as (

    select * from {{ ref('base_sl_web_issued_perk_code') }}

),

subscriptions_xf as (

    select
        subscriptions.customer_id,
        customers.user_id,
        subscriptions.id as subscription_id,
        subscriptions.created,  -- start of subscription (including trial if one existed)
        subscriptions.trial_start, -- start of trial (if existed)
        subscriptions.trial_end,  -- end of trial (if existed)
        subscriptions.current_period_start,  -- start of the current period (window for which charge applies)
        subscriptions.current_period_end,  -- end of the current period (window for which charge applies)
        subscriptions.canceled_at,  -- when the user canceled the subscription
        subscriptions.ended_at,  -- when the subscription actually ended 
        subscriptions.status,
        subscriptions.brand,
        subscriptions.interval,
        subscriptions.discount_id,
        subscriptions.is_deleted,
        subscriptions.promotion_id,
        subscriptions.metadata_migrated

    from subscriptions

    left join customers
        on subscriptions.customer_id = customers.customer_id

),

subscriptions_xf_2 as (

    select
        subscriptions_xf.user_id,
        subscriptions_xf.customer_id,
        subscriptions_xf.subscription_id,
        subscriptions_xf.created,  -- start of subscription (including trial if one existed)
        subscriptions_xf.trial_start, -- start of trial (if existed)
        subscriptions_xf.trial_end,  -- end of trial (if existed)
        subscriptions_xf.current_period_start,  -- start of the current period (window for which charge applies)
        subscriptions_xf.current_period_end,  -- end of the current period (window for which charge applies)
        subscriptions_xf.canceled_at,  -- when the user canceled the subscription
        subscriptions_xf.ended_at,  -- when the subscription actually ended 
        subscriptions_xf.status,
        subscriptions_xf.brand,
        subscriptions_xf.interval,
        subscriptions_xf.discount_id,
        subscriptions_xf.is_deleted,
        subscriptions_xf.promotion_id,
        subscriptions_xf.metadata_migrated,

        case  -- BEFORE L043 FIX
            when perks.perk is null then false
            else true
        end as perk_redeemed,

        perks.received_at as perk_received_at,
        perks.perk

    from subscriptions_xf

    left join perks
        on subscriptions_xf.user_id = perks.user_id
        and subscriptions_xf.created < perks.received_at  -- perk can only occur after trial/subscription starts
        and coalesce(subscriptions_xf.ended_at, current_date + 1) > perks.received_at  -- perk must be redeemed before trial/subscription ends, but can't be in future (i.e. tomorrow)

),

final as (

    select
        user_id,
        customer_id,
        subscription_id,
        created,  -- start of subscription (including trial if one existed)
        trial_start, -- start of trial (if existed)
        trial_end,  -- end of trial (if existed)

        case
            when trial_end is null then created
            when metadata_migrated is not null and "interval" = 'annual' then dateadd('year', -1, trial_end) -- when annual subscription migrated to Stripe 
            when metadata_migrated is not null and "interval" = 'month' then dateadd('month', -1, trial_end) -- when monthly subscription migrated to Stripe 
            else trial_end
        end as subscription_start,  -- start of PAID subscription (does not include trial period)

        current_period_start,  -- start of the current period (window for which charge applies)
        current_period_end,  -- end of the current period (window for which charge applies)
        canceled_at,  -- when the user canceled the subscription
        ended_at,  -- when the subscription actually ended 
        "status",
        brand,
        "interval",
        discount_id,
        is_deleted,
        promotion_id,
        metadata_migrated,
        perk_redeemed,
        perk_received_at,
        perk

    from subscriptions_xf_2

)

select * from final

```
</p>
</details>

<details>
<summary>Example dbt model AFTER running "sqlfluff fix"</summary>
<p>

```sql
-- filename: sl_stripe_subscriptions_with_perks_history.sql

{{
    config(
        materialized='table',
        sort=['user_id', 'created']
    )
}}

with subscriptions as (

    select * from {{ ref('stg_sl_stripe_subscriptions') }}

),

customers as (

    select * from {{ ref('stg_sl_stripe_customers') }}

),

perks as (

    select * from {{ ref('base_sl_web_issued_perk_code') }}

),

subscriptions_xf as (

    select
        subscriptions.customer_id,
        customers.user_id,
        subscriptions.id as subscription_id,
        subscriptions.created,  -- start of subscription (including trial if one existed)
        subscriptions.trial_start, -- start of trial (if existed)
        subscriptions.trial_end,  -- end of trial (if existed)
        subscriptions.current_period_start,  -- start of the current period (window for which charge applies)
        subscriptions.current_period_end,  -- end of the current period (window for which charge applies)
        subscriptions.canceled_at,  -- when the user canceled the subscription
        subscriptions.ended_at,  -- when the subscription actually ended
        subscriptions.status,
        subscriptions.brand,
        subscriptions.interval,
        subscriptions.discount_id,
        subscriptions.is_deleted,
        subscriptions.promotion_id,
        subscriptions.metadata_migrated

    from subscriptions

    left join customers
        on subscriptions.customer_id = customers.customer_id

),

subscriptions_xf_2 as (

    select
        subscriptions_xf.user_id,
        subscriptions_xf.customer_id,
        subscriptions_xf.subscription_id,
        subscriptions_xf.created,  -- start of subscription (including trial if one existed)
        subscriptions_xf.trial_start, -- start of trial (if existed)
        subscriptions_xf.trial_end,  -- end of trial (if existed)
        subscriptions_xf.current_period_start,  -- start of the current period (window for which charge applies)
        subscriptions_xf.current_period_end,  -- end of the current period (window for which charge applies)
        subscriptions_xf.canceled_at,  -- when the user canceled the subscription
        subscriptions_xf.ended_at,  -- when the subscription actually ended
        subscriptions_xf.status,
        subscriptions_xf.brand,
        subscriptions_xf.interval,
        subscriptions_xf.discount_id,
        subscriptions_xf.is_deleted,
        subscriptions_xf.promotion_id,
        subscriptions_xf.metadata_migrated,

        perks.perk is null, false) as perk_redeemed,  -- AFTER L043 FIX

        perks.received_at as perk_received_at,
        perks.perk

    from subscriptions_xf

    left join perks
        on subscriptions_xf.user_id = perks.user_id
            and subscriptions_xf.created < perks.received_at  -- perk can only occur after trial/subscription starts
            -- perk must be redeemed before trial/subscription ends, but can't be in future (i.e. tomorrow)
            and coalesce(subscriptions_xf.ended_at, current_date + 1) > perks.received_at

),

final as (

    select
        user_id,
        customer_id,
        subscription_id,
        created,  -- start of subscription (including trial if one existed)
        trial_start, -- start of trial (if existed)
        trial_end,  -- end of trial (if existed)

        case
            when trial_end is null then created
            -- when annual subscription migrated to Stripe
            when metadata_migrated is not null and "interval" = 'annual' then dateadd('year', -1, trial_end)
            -- when monthly subscription migrated to Stripe
            when metadata_migrated is not null and "interval" = 'month' then dateadd('month', -1, trial_end)
            else trial_end
        end as subscription_start,  -- start of PAID subscription (does not include trial period)

        current_period_start,  -- start of the current period (window for which charge applies)
        current_period_end,  -- end of the current period (window for which charge applies)
        canceled_at,  -- when the user canceled the subscription
        ended_at,  -- when the subscription actually ended
        "status",
        brand,
        "interval",
        discount_id,
        is_deleted,
        promotion_id,
        metadata_migrated,
        perk_redeemed,
        perk_received_at,
        perk

    from subscriptions_xf_2

)

select * from final

```
</p>
</details>

As shown in the second screenshot with git-diffs, this is producing an unexpected fix.

## Version
I have installed sqlfluff from source using `pip install git+https://github.com/sqlfluff/sqlfluff.git@master` (currently on commit: 9970bf2) so that I can take advantage of recent PRs (#868 and #964) instead of having to wait for the next release.

## Configuration

<details>
<summary>SQLFluff configuration</summary>
<p>
# For SQLFluff Rules reference, see:
# https://docs.sqlfluff.com/en/stable/rules.html#rules-reference
[sqlfluff]
verbose = 0
nocolor = False
dialect = postgres
templater = dbt
rules = None
exclude_rules = L032,L033,L034,L037,L044
recurse = 0
output_line_length = 120
runaway_limit = 10
ignore = parsing
ignore_templated_areas = True

[sqlfluff:indentation]
indented_joins = False
template_blocks_indent = True

# Some rules can be configured directly from the config common to other rules.
[sqlfluff:rules]
tab_space_size = 4
max_line_length = 120
indent_unit = space
comma_style = trailing
allow_scalar = True
single_table_references = consistent
only_aliases = True

# Some rules have their own specific config.
# All SQLFluff rules can be found at: https://docs.sqlfluff.com/en/stable/rules.html#rules-reference
# When a rule in not listed below, we inherit the default behavior from above.
[sqlfluff:rules:L003]
lint_templated_tokens = True

[sqlfluff:rules:L010]  # Keywords
capitalisation_policy = lower

[sqlfluff:rules:L014]  # Unquoted identifiers
extended_capitalisation_policy = lower
unquoted_identifiers_policy = all

[sqlfluff:rules:L016]
# Setting to True allows us to copy/paste long URLs as comments
ignore_comment_lines = True

[sqlfluff:rules:L030]  # Function names
capitalisation_policy = lower

[sqlfluff:rules:L038]
select_clause_trailing_comma = forbid

[sqlfluff:rules:L040]  # Null & Boolean Literals
capitalisation_policy = lower

[sqlfluff:rules:L042]
# By default, allow subqueries in from clauses, but not join clauses.
forbid_subquery_in = join

```
</p>
</details>

**Comment by barrywhart** [comment]
A question about "sometimes": can you tell if the behavior is consistent? That is, for a given input SQL, does SQLFluff **always** or **never** include the `coalesce()`? Also, could you add your succeeding and failing SQL examples as text comments? That will make it easier for someone to reproduce the issue.

**Comment by GClunies** [comment]
Sure thing @barrywhart, I am linting/fixing our entire dbt-repo at the moment so I will post the next occurrence when this happens (the others are already fixed manually and commited).

When I do run into it, I will `sqlfluff fix` multiples times to ensure the behaviour is consistent on the same input model.

**Comment by GClunies** [comment]
@barrywhart
1. Updated with a consistent example with SQL before and after fix.
2. Regarding "sometimes", if I use the same input SQL, I get a consistent unexpected fix. I've added this to the description.

**Comment by GClunies** [comment]
Here is an example of a model that returned the _expected_ `coalesce()` when fixing L043.

<details>
<summary>Example dbt model BEFORE running "sqlfluff fix"</summary>
<p>

```sql
--filename: staging_v_1_2_msw_itunes_trial_and_non_trial_terminations_by_brand_interval_date.sql

{{
    config(
        materialized = 'view'
    )
}}

/*
NOTE:
This model only contains non-trial terminations since MSW did not have a free
trial when we fetched data using schema version 1_1.
*/

with events_ref as (

    select
        *

    from {{ ref('staging_v_1_1_msw_itunes_subscription_event_schema_xf') }}
    
    where event in {{ itunes_termination_events_v_1_1() }}

),

final as (

    select
        event_date as "date",
        brand,
        "interval",
        'itunes'::text as termination_source_descriptor,
        quantity,

        case
            when event = 'Cancel' and consecutive_paid_periods = 0
                then true
            else false
        end as did_terminate_during_trial

    from events_ref

)

select * from final

```
</p>
</details>


<details>
<summary>Example dbt model BEFORE running "sqlfluff fix"</summary>
<p>

```sql
--filename: staging_v_1_2_msw_itunes_trial_and_non_trial_terminations_by_brand_interval_date.sql

{{
    config(
        materialized = 'view'
    )
}}

/*
NOTE:
This model only contains non-trial terminations since MSW did not have a free
trial when we fetched data using schema version 1_1.
*/

with events_ref as (

    select
        *

    from {{ ref('staging_v_1_1_msw_itunes_subscription_event_schema_xf') }}
    
    where event in {{ itunes_termination_events_v_1_1() }}

),

final as (

    select
        event_date as "date",
        brand,
        "interval",
        'itunes'::text as termination_source_descriptor,
        quantity,

        coalesce(event = 'Cancel' and consecutive_paid_periods = 0, false) as did_terminate_during_trial

    from events_ref

)

select * from final

```
</p>
</details>

**Comment by barrywhart** [comment]
I found a simpler test case to reproduce this issue. Simply run `sqlfluff fix test.sql --rules L043` on the following SQL file:
```
select
    subscriptions_xf.metadata_migrated,

    case  -- BEFORE L043 FIX
        when perks.perk is null then false
        else true
    end as perk_redeemed,

    perks.received_at as perk_received_at

from subscriptions_xf

```

**Comment by barrywhart** [comment]
Interestingly, blank lines seem to be significant. If I remove **any** of the three blank lines in the example above, the resulting SQL contains `not coalesce(`. But with those blank lines present, it does not.

**Comment by barrywhart** [comment]
I'm tinkering with this and may have found the issue. In the list of fixes below, note that the `case` is the target of **two** fixes:
* Delete
* Edit

This seems wrong, and would definitely explain why the `coalesce(` is missing from the final SQL. I'll do some more investigation to see if I can determine why.

```
(Pdb) pp fixes
[<LintFix: delete @[52](1, 4, 5) delete:'case'>,
 <LintFix: delete @[58](1, 4, 11) delete:'-- BEFORE L043 FIX'>,
 <LintFix: delete @[85](1, 5, 9) delete:'when'>,
 <LintFix: delete @[109](1, 5, 33) delete:'then'>,
 <LintFix: delete @[114](1, 5, 38) delete:'false'>,
 <LintFix: delete @[128](1, 6, 9) delete:'else'>,
 <LintFix: delete @[133](1, 6, 14) delete:'true'>,
 <LintFix: delete @[142](1, 7, 5) delete:'end'>,
 <LintFix: edit @[52](1, 4, 5) edt:'case'->'not coalesce('>,
 <LintFix: edit @[108](1, 5, 32) edt:' '->', false)'>]
```
